Help Center: Update gpt answer flow, and add A/B test

yoavf opened the pull request:
As a follow-up to #75838, this makes some UX changes to the GPT answer generation and enables the experiment to test GPT-generated answers against our current questions and links deflection method 

## Proposed Changes

* Make the GPT answer loading non-blocking; Add text setting expectations on how long loading will take and informing the user they can submit their request without waiting.
* Add A/B test assignment loading in `HelpCenterContactPage` and pass it via the route links param to `HelpCenterContactForm`


A/B test description: pbxNRc-2Ag-p2
Abacus link: 21214-explat-experiment


## Testing Instructions
- Assign yourself to either the control or treatment variant
- In `control`, the process or submitting a support request through the help center should be unchanged
- In `treatment` , you will be presented with a GPT-generated answer instead of the question and links format before the final submission. 

### Control
![control](https://github.com/Automattic/wp-calypso/assets/844866/d7b137da-714c-41c7-a16a-94d9f9a0ba36)

### Treatment
![treatment](https://github.com/Automattic/wp-calypso/assets/844866/70ca4625-efe7-44ba-9c10-38a220f9a42c)


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

yoavf committed:
add experiment

yoavf committed:
keep adding declarations until the tests passes

Can't say I get the point of that

jessie-ross commented on packages/help-center/src/components/index.d.ts:
```diff
@@ -147,3 +147,17 @@ declare module '@automattic/state-utils' {
 
 declare module 'calypso/lib/mobile-app';
+
+declare module 'calypso/lib/explat' {
```
Hmm, I'm not sure why these types are necessary, the explat client library is fully typed.

yoavf replied:
Yeah, I'm not sure either; just a shot in the dark since I was getting unit-tests failing without them.

yoavf replied:
This is what I was getting [#](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/10013369?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=10013369_6843_449.6841&logFilter=debug&logView=flowAware):
```
....STDOUT....

[09:28:10 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/10013369?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=10013369_6844_6844&logFilter=debug&logView=flowAware)
    client/lib/explat/internals/log-error.ts(1,19): error TS7016: Could not find a declaration file for module 'calypso/lib/wp'. '/home/teamcity-1/buildAgent/work/c4a9d5b38c1dacde/client/lib/wp/node.js' implicitly has an 'any' type.

[09:28:10 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/10013369?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=10013369_6845_6845&logFilter=debug&logView=flowAware)
      Try `npm i --save-dev @types/wp` if it exists or add a new declaration (.d.ts) file containing `declare module 'calypso/lib/wp';`

[09:28:10 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/10013369?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=10013369_6846_6846&logFilter=debug&logView=flowAware)
    client/lib/explat/internals/log-error.ts(2,27): error TS7016: Could not find a declaration file for module 'calypso/server/lib/logger'. '/home/teamcity-1/buildAgent/work/c4a9d5b38c1dacde/client/server/lib/logger/index.js' implicitly has an 'any' type.

[09:28:10 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/10013369?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=10013369_6847_6847&logFilter=debug&logView=flowAware)
      Try `npm i --save-dev @types/calypso` if it exists or add a new declaration (.d.ts) file containing `declare module 'calypso/server/lib/logger';`

[09:28:10 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/10013369?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=10013369_6848_6848&logFilter=debug&logView=flowAware)
    client/lib/explat/internals/fetch-experiment-assignment.ts(1,19): error TS7016: Could not find a declaration file for module 'calypso/lib/wp'. '/home/teamcity-1/buildAgent/work/c4a9d5b38c1dacde/client/lib/wp/node.js' implicitly has an 'any' type.

[09:28:10 ](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/10013369?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=10013369_6849_6849&logFilter=debug&logView=flowAware)
      Try `npm i --save-dev @types/wp` if it exists or add a new declaration (.d.ts) file containing `declare module 'calypso/lib/wp';`
      ```

yoavf replied:
Ok, I don't know what I'm doing here :) @alshakero can you help me understand what I need to do here?

<details><summary>These are the errors I get if I run `npm run build` in the help-center package directory:</summary>
<p>

```
../../client/lib/explat/internals/log-error.ts:1:19 - error TS7016: Could not find a declaration file for module 'calypso/lib/wp'. '/Users/yoavf/dev/wp-calypso/client/lib/wp/node.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/wp` if it exists or add a new declaration (.d.ts) file containing `declare module 'calypso/lib/wp';`

1 import wpcom from 'calypso/lib/wp';
                    ~~~~~~~~~~~~~~~~

../../client/lib/explat/internals/log-error.ts:2:27 - error TS7016: Could not find a declaration file for module 'calypso/server/lib/logger'. '/Users/yoavf/dev/wp-calypso/client/server/lib/logger/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/calypso` if it exists or add a new declaration (.d.ts) file containing `declare module 'calypso/server/lib/logger';`

2 import { getLogger } from 'calypso/server/lib/logger';
                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~

../../client/lib/explat/internals/fetch-experiment-assignment.ts:1:19 - error TS7016: Could not find a declaration file for module 'calypso/lib/wp'. '/Users/yoavf/dev/wp-calypso/client/lib/wp/node.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/wp` if it exists or add a new declaration (.d.ts) file containing `declare module 'calypso/lib/wp';`

1 import wpcom from 'calypso/lib/wp';
                    ~~~~~~~~~~~~~~~~

../../client/lib/explat/internals/log-error.ts:1:19 - error TS7016: Could not find a declaration file for module 'calypso/lib/wp'. '/Users/yoavf/dev/wp-calypso/client/lib/wp/node.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/wp` if it exists or add a new declaration (.d.ts) file containing `declare module 'calypso/lib/wp';`

1 import wpcom from 'calypso/lib/wp';
                    ~~~~~~~~~~~~~~~~

../../client/lib/explat/internals/log-error.ts:2:27 - error TS7016: Could not find a declaration file for module 'calypso/server/lib/logger'. '/Users/yoavf/dev/wp-calypso/client/server/lib/logger/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/calypso` if it exists or add a new declaration (.d.ts) file containing `declare module 'calypso/server/lib/logger';`

2 import { getLogger } from 'calypso/server/lib/logger';
                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~

../../client/lib/explat/internals/fetch-experiment-assignment.ts:1:19 - error TS7016: Could not find a declaration file for module 'calypso/lib/wp'. '/Users/yoavf/dev/wp-calypso/client/lib/wp/node.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/wp` if it exists or add a new declaration (.d.ts) file containing `declare module 'calypso/lib/wp';`

1 import wpcom from 'calypso/lib/wp';
                    ~~~~~~~~~~~~~~~~
```


</p>
</details>

alshakero replied:
I suppose the lib itself is typed, but its instance in Calypso depends on untyped dependencies.

Namely this: https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/explat/internals/fetch-experiment-assignment.ts#L1
And this: https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/explat/internals/log-error.ts#L2

Typing these will probably cut it. Although, we should avoid the Calypso instance altogether in Atomic.

jessie-ross reviewed (changes requested):
It looks like when the experiment itself is loading, we display the control experience. For an unloaded experiment we need to display neither control or treatment until the experiment loads - we'd need to display a loading indicator for both experiences where the AI is displayed.

yoavf committed:
try removing type defs again

yoavf committed:
move the experiment assignment earlier so that we don't show the control variant while waiting for it to load

yoavf commented:
> It looks like when the experiment itself is loading, we display the control experience. For an unloaded experiment we need to display neither control or treatment until the experiment loads - we'd need to display a loading indicator for both experiences where the AI is displayed.

Good catch. I've moved the experiment assignment earlier to when we load the different contact mode options (so the loading indicator for that will also wait for the experiment to be loaded).

yoavf committed:
Change behaviour for the GPT experiment:

- Remove one AI reference
- Add text setting expectation on how long loading will take, and letting the user know they can submit their message right away
- Display a reminder of the user message just before the submit button
- The CTA button will be enabled immediately
- Do not display the links retrieved by GPT - they're less useful and having them displayed requires more complex UI changes (like changing the button stickyness)

yoavf committed:
minor design fix

yoavf committed:
enable the feature in production along with the experiment

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7988967

Thank you @yoavf for including a screenshot in the description! This is really helpful for our translators.

alshakero commented on config/production.json:
```diff
@@ -47,7 +47,7 @@
 		"happychat": true,
 		"help": true,
-		"help/gpt-response": false,
+		"help/gpt-response": true,
```
Copying this here: https://github.com/Automattic/wp-calypso/pull/75838/files#r1187604756

yoavf replied:
Yep, understood. We'll test just in caplyso first.

alshakero commented on packages/help-center/src/components/help-center-contact-form.tsx:
```diff
@@ -444,17 +443,17 @@ export const HelpCenterContactForm = () => {
 	};
 
-	const { isFetching: isFetchingUrls, data: links } = useJetpackSearchAIQuery(
+	const { isFetching: isFetchingGPTUrls, data: links } = useJetpackSearchAIQuery(
```
Optional: `useJetpackSearchAIQuery` should accept `enabled` param and forward it to `useQuery`. That would be more accurate than `shouldLoadGptAnswer ? debouncedMessage : ''`

yoavf replied:
You're right, that's much better. I went a step further and made useJetpackSearchAIQuery to have a single pre-defined config param, which I think is more readable.

alshakero commented on packages/help-center/src/components/help-center-contact-form.tsx:
```diff
@@ -444,17 +443,17 @@ export const HelpCenterContactForm = () => {
 		'urls'
 	);
-	const { isFetching: isFetchingResponse, data: gptResponse } = useJetpackSearchAIQuery(
+	const { isFetching: isFetchingGPTAnswer, data: gptResponse } = useJetpackSearchAIQuery(
```
+1

alshakero commented on packages/help-center/src/components/help-center-contact-page.tsx:
```diff
@@ -16,6 +16,7 @@ import classnames from 'classnames';
 import { FC } from 'react';
 import { useSelector } from 'react-redux';
 import { Link, LinkProps } from 'react-router-dom';
+import { useExperiment } from 'calypso/lib/explat';
```
You're starting to enter the Help Center inferno 🚒 

I'd be a bit cautious with Calypso imports inside `packages`. Especially ones that do network requests.

1. They often pull plenty of Calypso bytes into the editor. And our editor is already slow.
2. Because the Help Center runs on Atomic, and most of these requests don't work on Atomic. I know this one doesn't because it uses [wpcom.req](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/explat/internals/fetch-experiment-assignment.ts#L14) which blocks requests from Atomic.

cc @jessie-ross 👋🏼 I wonder if you have any advice on this. Is there an easy way to run ExPlat on Atomic?

yoavf replied:
Could I use [explat-client](https://github.com/Automattic/wp-calypso/blob/2c1818291dc4d7758f73f6f444df5bbcd6f0aebf/packages/explat-client/README.md#L11) directly in this case?

alshakero replied:
Yes, that's the way, but _I think_ you'll still need to initiate an instance like [so](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/explat/index.ts#L11).

```js
const exPlatClient = createExPlatClient( {
	fetchExperimentAssignment,
	getAnonId,
	logError,
	isDevelopmentMode,
} );
``` 

And `fetchExperimentAssignment` depends on `wpcom.req`, which [requests](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/explat/internals/fetch-experiment-assignment.ts#L21) from  `'/experiments/0.1.0/assignments/calypso'` endpoint. 

This endpoint needs to be public (un-authed) and allows CORS, in which case you can just use standard `fetch`.

And guess what? I just checked and it is! I suppose you just need to pass the user ID as `anon_id` query param. @jessie-ross knows the answer.


so your `fetchExperimentAssignment` would be

```js
export default function fetchExperimentAssignment( {
	experimentName,
	userId,
}: {
	experimentName: string;
	userId: number | string | null;
} ): Promise< unknown > {
    const query = new URLSearchParams();
	query.add('experiment_name', experimentName);
    query.add('anon_id', userId);

	return fetch('https://public-api.wordpress.com/wpcom/v2/experiments/0.1.0/assignments/calypso' + query.toString())
}
```

jessie-ross replied:
> Is there an easy way to run ExPlat on Atomic?

Ah, I see, I don't think we've run an experiment on atomic yet.

Will this code will always run on atomic, or will it also sometime run in simple?

If it the latter we'll need a dynamic approach that switches between techniques.

If it is the former, we could create an atomic client just as you mention.

> This endpoint needs to be public (un-authed) and allows CORS, in which case you can just use standard fetch.

> And guess what? I just checked and it is! I suppose you just need to pass the user ID as anon_id query param. @jessie-ross knows the answer.

This will likely work, but requests should ideally be authenticated - it provides extra checks, safety and would allow us to measure predicted churn, is there really no way to call the wp-api from atomic?

alshakero replied:
There sure is! You just need to create a proxy endpoint in Jetpack or ETK. Something like [this](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/help-center/class-wp-rest-help-center-search.php#L64). 

This makes the request go through the site (same domain), `yoursite.wordpress.com/wp-json/blabla`, then to Jetpack (which signs it on the user behalf), then to public-api. Then, on the client, you can do [this](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/support-queries/use-sibyl-query.ts#L18-L29).

yoavf replied:
@alshakero Do you see any problem in using `canAccessWpcomApis`/`wpcomRequest`/`apiFetch` in `fetchExperimentAssignment`

Something like 
```js
async function fetchExperimentAssignment( {
	experimentName,
	userId,
}: {
	experimentName: string;
	userId: number | string | null;
} ): Promise< unknown > {
	return canAccessWpcomApis()
		? wpcomRequest( {
				path: `experiments/0.1.0/assignments/calypso/?experiment_name=${ experimentName }&anonId=${
					userId ?? undefined
				}`,
				apiNamespace: 'wpcom/v2/',
				apiVersion: '2',
		  } )
		: apiFetch( {
				global: true,
				path: `experiments/0.1.0/assignments/calypso/?experiment_name=${ experimentName }&anonId=${
					userId ?? undefined
				}`,
		  } as APIFetchOptions );
}
```

jessie-ross replied:
@yoavf Do you have the capacity to add the proxy endpoint as part of this work?

alshakero replied:
@yoavf that works and I certainly don't want to block you, but I'd just double-check that this doesn't pull a ton of Calypso bytes into the Help Center. You can check by running `yarn build` in `apps/editing-toolkit`, then check the bundle size `apps/editing-toolkit/editing-toolkit-plugin/help-center/dist/help-center.min.js`. 

For comparision, you can pull a built ETK from `trunk` of any other PR, it's available under 

![image](https://github.com/Automattic/wp-calypso/assets/17054134/52bbaddf-eb65-4f5d-9180-3b03ead14470)

Then

![image](https://github.com/Automattic/wp-calypso/assets/17054134/5cd8d1bc-3a3c-407c-bdca-9c179629b206)

Currently, it's `976 388 bytes`

Also, I don't know how important auth is for ExPlat.

alshakero commented on packages/help-center/src/components/help-center-contact-page.tsx:
```diff
@@ -48,7 +49,20 @@ export const HelpCenterContactPage: FC = () => {
+		{ isEligible: featureEnabledGPTResponse }
+	);
+
+	const loadGpt =
+		! isLoadingExperimentAssignment && experimentAssignment?.variationName === 'treatment';
```
I'd wrap this with a `Boolean()`, it may return `undefined` on network failure; when `isLoadingExperimentAssignment` is false and `experimentAssignment` is `undefined`.

alshakero reviewed (changes requested):
Pretty clean! Left some comments.

yoavf committed:
Only apply the "busy" status to the CTA button when loading the GPT links (<4 seconds).
This is more inline with the "disabled" status in the Sibyl version, where the button is disabled until the links are loaded.

yoavf committed:
wrap in Boolean() to prevent unexpected 'undefined'

yoavf committed:
Refactor useJetpackSearchAIQuery to use a config object as a param

Apply to callers, and use the new enabled proprety instead of passing an empty query string

jessie-ross commented on packages/help-center/src/components/help-center-contact-page.tsx:
```diff
@@ -142,7 +157,7 @@ export const HelpCenterContactPage: FC = () => {
 
 				<div className={ classnames( 'help-center-contact-page__boxes' ) }>
-					<Link to="/contact-form?mode=FORUM">
+					<Link to={ `/contact-form?mode=FORUM&loadGpt=${ loadGpt.toString() }` }>
```
Is it possible for this link to be clicked before `isLoading` is true? If so we'd be in the same place where we'd show the control experience while loading.

yoavf replied:
Nope, line 123 above has an early return.

jessie-ross replied:
Ah good, then from the experimentation side of things we're good (once the client is configured for atomic.)

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7988967).
